README.md
idoit
=====

[![CI](https://github.com/nodeca/idoit/workflows/CI/badge.svg)](https://github.com/nodeca/idoit/actions)
[![NPM version](https://img.shields.io/npm/v/idoit.svg?style=flat)](https://www.npmjs.org/package/idoit)
[![Coverage Status](https://coveralls.io/repos/github/nodeca/idoit/badge.svg?branch=master)](https://coveralls.io/github/nodeca/idoit?branch=master)


> Redis-backed task queue engine with advanced task control and eventual consistency.

- Task grouping, chaining, iterators for huge ranges.
- Postponed & scheduled task run.
- Load distribution + worker pools.
- Easy to embed.


Features in details
-------------------

`idoit` provides advanced control to implement so

**Grouping**. Special `group` task execute children tasks and wait until all
complete. Useful for map/reduce logic.

**Chaining**. Special `chain` task execute children one-by-one. Also useful for
map-reduce or splitting very complicated tasks to more simple steps.

**Mapping iterator**. Special feature for huge payloads, to produce chunks on
demand. Benefits:

- No lags on mapping phase, chunks processing starts immediately.
- Easy to optimize DB queries to build chunks of equal size
  (skip + limit queries are very slow on huge data).

**Progress**. When you use groups/chain/map scenarios, it's easy to monitor
total progress via top parent. Long standalone tasks also can notify user about
progress change.

**Worker pools**. You can split tasks by different processes. For example,
if you don't wish heavy tasks to block light ones.

**Sheduler**. Built-in cron allows to execute tasks on given schedule.


## Data consistency

- All data in redis are evertually consistent.
- Task can not be lost, but CAN run twice on edge cases (if process crash when
  task function was about to finish)
- Progress can count "faster" if `task.progressAdd()` used and process crash
  before task complete. But that's not critical, since such info can be used
  only for interface progress bars updates. In most cases you will not see the
  difference.


Install
-------

`node.js` 6+ and `redis` 3.0+ required.

```sh
npm install idoit --save
```


API
---

### new Queue({ redisURL, concurrency = 100, name = 'default', ns = 'idoit:' })

 - **redisURL** (String) - redis connection url.
 - **concurrency** (Number) - max tasks to consume in parallel
   by single worker, 100 by default.
 - **pool** (String) - worker pool name, "default" if not set. Used if
   this queue instance consumes tasks only (after `.start()`). You
   can route tasks to specific pools of workers to avoid unwanted locks.
   You can set `pool` to Array, `[ 'pool1', 'pool2' ]` to consume tasks from
   several pools (for development/testing purposes).
 - **ns** (String) - data namespace, currently used as redis keys prefix,
   "idoitqueue:" by default.

It's a good practice to have separate worker pools for heavy blocking tasks and
non-blocking ones. For example, nobody should block sending urgent emails. So,
create several worker processes, pin those to different pools and set proper
tasks concurrency. Non-blocking tasks can be cunsumed in parallel, and you can
be ok with default `concurrency` = 100. Blocking tasks should be consumed
one-by-one, set `concurrency` = 1 for those workers.

__Note.__ It may happen, that you remove some task types from your app. In this
case orphaned data will be wiped after 3 days.


### .registerTask(options), .registerTask(name [, cron], process)

Options:

 - **name** (String) - the task's name.
 - **baseClass** (Function) - optional, base task's constructor, "Task" by default.
 - **init** (Function) - optional, used for async task initialization, should return `Promise`
   - **this** (Object) - current task (task total is available as `this.total`).
 - **taskID** (Function) - optional, should return new task id. Needed only for
   creating "exclusive" tasks, return random value by default, called as:
   `function (taskData)`. Sugar: if you pass plain string, it will be wrapped to
   function, that always return this string.
 - **process** (Function) - main task function, called as: `task.process(...args)`. Should return `Promise`
   - **this** (Object) - current task.
 - **retry** (Number) - optional, number of retry on error, default 2.
 - **retryDelay** (Number) - optional, delay in ms after retries, default 60000 ms.
 - **timeout** (Number) - optional, execution timeout, default 120000 ms.
 - **total** (Number) - optional, max progress value, default 1. If you don't
   modify behaviour progress starts with 0 and become 1 on task end.
 - **postponeDelay** (Number) - optional, if postpone is called without delay,
   delay is assumed to be equal to this value (in milliseconds).
 - **cron** (String) - optional, cron string ("15 */6 * * *"), default null.
 - **track** (Number) - default 3600000ms (1hr). Time to remember scheduled
   tasks from cron to avoid rerun if several servers in cluster have wrong
   clocks. Don't set too high for very frequent tasks, because it can occupy
   a lot of memory.

### .getTask(id)

Get task by id. Returns a Promise resolved with task or with `null` if task not exist.

__Task fields you can use:__

- **total** - total task progress
- **progress** - current task progress
- **result** - the task result
- **error** - the task error


### .cancel(id)

Cancel task. Returns a Promise resolved with task.

__Note.__ You can cancel only tasks without parent.


### .start()

Start worker and begin task data consume. Return `Promise`, resolved when queue
is ready (call `.ready()` inside).

If `pool` was specified in cunstructor, only tasks routed to this pull will
be consumed.


### .shutdown()

Stop accepting new tasks from queue. Return `Promise`, resolved when all active
tasks in this worker complete.


### .ready()

Return `Promise`, resolved when queue is ready to operate (after 'connect'
event, see below).


### .options(opts)

Update constructor options, except redisURL.


### .on('eventName', handler)

`idoit` is an `EventEmitter` instance that fires some events:

- `ready` when redis connection is up and commands can be executed
  (tasks can be registered without connection)
- `error` when an error has occured.
- `task:progress`, `task:progress:<task_id>` - when task update progress.
  Event data is: { id, uid, total, progress }
- `task:end`, `task:end:<task_id>` - when task end. Event data is: { id, uid }


### .\<taskName\>(...params)

Create new Task with optional params.


### task.options({ ... })

Override task properties. For example, you may wish to assign specific
group/chain tasks to another pool.


### task.run()

Run task immediately. Returns a Promise resolved with task id.


### task.postpone([delay])

Postpone task execution to `delay` milliseconds (or to `task.postponeDelay`).

Returns a Promise resolved with task id.


### task.restart([add_retry] [, delay])

Restart currently running task.

- **add_retry** (Boolean) - optional, whether to increase retry count or not
  (default: false)
  - if `true`, retry count is increased, and task doesn't get restarted in case
    it's exceeded
  - if `false`, retry count stays the same, so a task can restart itself
    indefinitely
- **delay** (Number) delay before restart in milliseconds
  (default: `task.retryDelay`).

Note, `idoit` already has built-in restart logic on task errors. Probably,
you should not use this method directly. It's exposed for very specific cases.


### task.progressAdd(value)

Increment current task progress.

Returns a Promise resolved with task id.


### task.setDeadline(timeLeft)

Update current task deadline.

Returns a Promise resolved with task id.


Special tasks
-------------

### group

Create a new task, executing children in parallel.

```javascript
queue.group([
  queue.children1(),
  queue.children2(),
  queue.children3()
]).run()
```

Group result is unsorted array of children result.

### chain

Create a new task, executing children in series. If any of children fails -
chain fails too.

```javascript
queue.registerTask('multiply', (a, b) => a * b);
queue.registerTask('subtract', (a, b) => a - b);

queue.chain([
  queue.multiply(2, 3), // 2 * 3 = 6
  queue.subtract(10),   // 10 - 6 = 4
  queue.multiply(3)     // 3 * 4 = 12
]).run()
```

Result of previous task pass as last argument of next task. Result of chain
is result of last task in chain.

### iterator

A special way to run huge mapping in lazy style (on demand). See comments below.

```javascript
// register iterator task
queue.registerTask({
  name: 'lazy_mapper',
  baseClass: Queue.Iterator,
  // This method is called on task begin and on every child end. It can be
  // a generator function or function that return `Promise`.
  * iterate(state) {
    // ...

    // Three types of output states possible: ended, do nothing & new data.
    //
    // 1. `null` - end reached, iterator should not be called anymore.
    // 2. `{}`   - idle, there are enougth subtasks in queue, try to call
    //             iterator later (when next child finish).
    // 3. {
    //      state    - new iterator state to remember (for example, offset for
    //                 db query), any serializeable data
    //      tasks    - array of new subtasks to push into queue
    //    }
    //
    // IMPORTANT! Iterator can be called in parallel from different workers. We
    // use input `state` to resolve collisions on redis update. So, if you
    // create new subtasks:
    //
    // 1. new `state` MUST be different (for all previous states)
    // 2. `tasks` array MUST NOT be empty.
    //
    // In other case you should signal about 'end' or 'idle'.
    //
    // Invalid combination will cause 'end' + error event.
    //
    return {
      state: newState,
      tasks: chunksArray
    };
  }
});

// run iterator
queue.lazy_mapper().run();
```

Why this crazy magic was invented?

Imagine that you need to rebuild 10 millions of forum posts. You wish to split
work by equal small chunks, but posts have no sequential integer enumeration,
only mongo IDs. What can you do?

- Direct `skip` + `limit` requests are very expensive on big collections in any
  database.
- You can not split by date intervals, because posts density varies a lot from
  first to last post.
- You can add an indexed field with random number to every post. Then split by
  intervals. That will work, but will cause random disk access - not cool.

Solution is to use iterative mapper, wich can remember "previous position". In
this case, you will do `range` + `limit` requests instead of `skip` + `limit`.
That works well with databases. Additional bonuses are:

- You do not need to keep all subtasks in queue. For example, you can create
  100 chunks and add next 100 when previous are about to finish.
- Mapping phase become distributed and you can start monitoring total progress
  immediately.


Dev notes
---------

Quik-run redis via docker:

```sh
# start
docker run -d -p 6379:6379 --name redis1 redis
# stop
docker stop redis1
docker rm redis1
```


Why one more queue?
-------------------

Of cause, we are familiar with [kue](https://github.com/Automattic/kue),
[celery](http://www.celeryproject.org/) and [akka](http://akka.io/).
Our target was have a balance between simplicity and power. So, we don't know
if `idoit` works well in cluster with thousands of instances. But it should be ok
in smaller volumes and it's really easy to use.

[kue](https://github.com/Automattic/kue) was not ok for our needs, because:

- it's "priorities" concept is not flexible and does not protect well from
  locks by heavy tasks
- no task grouping/chaining and so on
- no strong guarantees of data consistency

In `idoit` we cared about:

- task group/chain operations & pass data between tasks (similar to celery)
- worker pools to isolate task execution by types.
- easy to use & install (only redis needed, can run in existing process)
- eventual consistency of stored data
- essential sugar like built-in scheduler
- iterative mapper for huge payloads (unique feature, very useful for many
  maintenance tasks)
- task progress tracking
- avoid global locks

Redis still can be a point of failure, but that's acceptable price for
simplicity. Of cause you can get a better availability via distributed message
buses like RMQ. But in many cases it's more important to keep things simple.
With `idoit` you can reuse existing technologies without additional expences.


CHANGELOG.md
4.0.0 / 2021-02-22
------------------

- Node.js 14.+ required.
- Refactored iterator to avoid custom scripts use.
- Moved transaction script to separate package.
- Code cleanup (replace `Promise` with `async`).


3.0.0 / 2020-07-29
------------------

- Node 10.+ required.
- Deps bump.
- Use es6 classes where possible.
- Make `.extend()` work in inherited classes.


2.1.0 / 2020-03-17
------------------

- Deps bump.
- Improve tests.


2.0.0 / 2017-07-04
------------------

- Switch to native async/await. Now requires node v7.+ to run.


1.2.3 / 2017-07-04
------------------

- Fix Queue.cancel crash when some child tasks are already finished.


1.2.2 / 2017-04-17
------------------

- Simplify children init from chain/group inherited tasks.


1.2.1 / 2017-03-29
------------------

- Fix empty result store in chains & groups.


1.2.0 / 2017-03-16
------------------

- Add API method to force task restart, #2.
- Implemented active chunks tracking for iterators.


1.1.1 / 2017-03-03
------------------

- Quick-fix to suppress events from iterator's children when
  parent was canceled.


1.1.0 / 2017-02-27
------------------

- Simplify extending group/chain task via `.init()` method override.


1.0.1 / 2017-02-07
------------------

- Fix `task:end` event emit on task cancel (for nested tasks).


1.0.0 / 2016-11-04
------------------

- First release.


lib/task_template.js
'use strict';


const utils   = require('./utils');
const Command = require('./command');


function TaskTemplate(queue, ...args) {
  this.queue = queue;

  this.uid      = utils.random();
  this.retries  = 0;
  this.total    = 1;
  this.progress = 0;
  this.state    = 'waiting';
  this.args     = args;

  // Fields for `.toObject()/.fromObject()`
  this.__serializableFields = [
    'name',
    'id',
    'pool',
    'uid',
    'retries',
    'total',
    'progress',
    'state',
    'parent',
    'parent_uid',
    'parent_pool',
    'args',
    'result',
    'error',
    // keep it to access without task class cast on cancel/error
    'removeDelay'
  ];
}


TaskTemplate.prototype.name         = 'task';
TaskTemplate.prototype.pool         = 'default';
TaskTemplate.prototype.retry        = 2;
TaskTemplate.prototype.retryDelay   = 60000;
TaskTemplate.prototype.timeout      = 120000;
TaskTemplate.prototype.removeDelay  = 3 * 24 * 60 * 60 * 1000;
TaskTemplate.prototype.__prepare_promise__ = null;


// Update task options
//
TaskTemplate.prototype.options = function (opts) {
  Object.assign(this, opts);
  return this;
};


// (internal) Prepare task prior to run. Used to modify templates
// behaviour on inherit.
//
// !!! Don't touch this method, override `.init()` to extend
// registered tasks.
//
TaskTemplate.prototype.prepare = async function () {
  this.id = this.taskID(...this.args);

  // .init() can be simple sync function
  await this.init();
};


// Override to customize task init. Async functions should return `Promise`
//
TaskTemplate.prototype.init = function () {};


// Create random task ID
//
TaskTemplate.prototype.taskID = function () {
  return utils.random();
};


// Put task to queue and run immediately
//
TaskTemplate.prototype.run = async function () {
  await this;
  return this.queue.__addTask__(this);
};


// Postpone task to `delay` ms.
//
TaskTemplate.prototype.postpone = async function (delay = this.postponeDelay) {
  await this;
  return this.queue.__addTask__(this, delay);
};


// Force restart of a running task
//
TaskTemplate.prototype.restart = async function (add_retry, delay) {
  if (typeof add_retry === 'number') {
    // restart(delay)
    delay = add_retry;
    add_retry = false;
  }

  if (typeof add_retry === 'undefined') add_retry = false;
  if (typeof delay === 'undefined') delay = this.retryDelay;

  await this;
  return this.queue.__restartTask__(this, add_retry, delay);
};


// Update task deadline, set to `timeLeft` ms.
//
TaskTemplate.prototype.setDeadline = async function (timeLeft) {
  let prefix = this.queue.__prefix__;
  let time = utils.redisToMs(await this.queue.__redis__.time());

  let success = await this.queue.__redis__.transaction(
    JSON.stringify({
      if: [
        [ String(this.__deadline__), [ 'zscore', `${prefix}locked`, this.id ] ]
      ],
      exec: [
        [ 'zadd', `${prefix}locked`, timeLeft + time, this.id ]
      ]
    })
  );

  if (success) {
    this.__deadline__ = timeLeft + time;
  }
};


// Add `value` to task progress, and propagate update to parent
//
TaskTemplate.prototype.progressAdd = async function (value) {
  // TODO: that's not safe on parallel run, but nobody cares...
  // Progress value has no strict requirements.

  let time   = utils.redisToMs(await this.queue.__redis__.time());
  let prefix = this.queue.__prefix__;

  let transaction = {
    if: [
      [ 1, [ 'exists', `${prefix}${this.id}` ] ]
    ],
    exec: [
      [ 'hincrby', `${prefix}${this.id}`, 'progress', value ]
    ]
  };

  if (this.parent) {
    transaction.exec.push([ 'zadd', `${prefix}${this.parent_pool}:commands`, time, Command.fromObject({
      to:     this.parent,
      to_uid: this.parent_uid,
      type:   'progress',
      data:   value
    }).toJSON() ]);
  }

  let success = await this.queue.__redis__.transaction(
    JSON.stringify(transaction)
  );

  if (success) {
    let progress = await this.queue.__redis__.hget(`${prefix}${this.id}`, 'progress');

    let eventData = {
      id: this.id,
      uid: this.uid,
      total: this.total,
      progress: +progress
    };

    this.queue.emit('task:progress', eventData);
    this.queue.emit(`task:progress:${this.id}`, eventData);
  }
};


// Handle command from other tasks. Route to `handleCommand_<type>` methods
//
TaskTemplate.prototype.handleCommand = async function (command) {
  let handlerName = `handleCommand_${command.type}`;

  // Unknown command? -> ignore & send notification.
  // That should never happen.
  if (!this[handlerName]) {
    this.queue.emit(
      'error',
      new Error(`idoit error: ignored unknown command for "${this.name}" - ${command.toJSON()}`)
    );
    return;
  }

  return this[handlerName](command);
};


// Process increment progress cmd (used in group, chain and iterator)
//
TaskTemplate.prototype.handleCommand_progress = async function (command) {
  let time   = utils.redisToMs(await this.queue.__redis__.time());
  let prefix = this.queue.__prefix__;
  let query  = this.queue.__redis__.multi();

  // Increment progress if task not finished
  let transaction = {
    if: [
      [ 1, '==', [ 'zrem', `${prefix}${this.pool}:commands_locked`, command.toJSON() ] ],
      [ JSON.stringify('finished'), '!=', [ 'hget', `${prefix}${this.id}`, 'state' ] ]
    ],
    exec: [
      [ 'hincrby', `${prefix}${this.id}`, 'progress', command.data ]
    ]
  };

  query.transaction(JSON.stringify(transaction));

  if (this.parent) {
    // Send command to parent (even if transaction rejected)
    query.zadd(`${prefix}${this.parent_pool}:commands`, time, Command.fromObject({
      to:     this.parent,
      to_uid: this.parent_uid,
      type:   'progress',
      data:   command.data
    }).toJSON());
  }

  let res = await query.exec();

  if (res[0][1]) {
    let eventData = {
      id: this.id,
      uid: this.uid,
      total: this.total,
      progress: this.progress + command.data
    };

    this.queue.emit('task:progress', eventData);
    this.queue.emit(`task:progress:${this.id}`, eventData);
  }
};


// Process `activate` command (valid in `waiting` state only)
//
TaskTemplate.prototype.handleCommand_activate = async function (command) {
  let prefix      = this.queue.__prefix__;
  let prefix_pool = `${this.queue.__prefix__}${this.pool}:`;

  await this.queue.__redis__.transaction(
    JSON.stringify({
      if: [
        [ 1, [ 'zrem', `${prefix_pool}commands_locked`, command.toJSON() ] ],
        [ JSON.stringify('waiting'), [ 'hget', `${prefix}${this.id}`, 'state' ] ]
      ],
      exec: [
        [ 'srem', `${prefix}waiting`, this.id ],
        [ 'sadd', `${prefix_pool}startable`, this.id ],
        [ 'hset', `${prefix}${this.id}`, 'state', JSON.stringify('startable') ]
      ]
    })
  );
};


// Error command handler (used in group, chain and iterator)
//
TaskTemplate.prototype.handleCommand_error = async function (command) {
  let prefix = this.queue.__prefix__;
  let time   = utils.redisToMs(await this.queue.__redis__.time());


  let tasks = await this.__fillCancelInfo__();
  let cancelIDs = tasks.map(t => t.id).filter(id => id !== this.id);

  // Fetch children data to get pools and tasks `removeDelay`
  //
  // TODO: This make task dependent on registered classes in this
  // queue instance. Consider rework
  //
  let rawTasks = await this.queue.__getRawTasks__(cancelIDs);

  let pools = rawTasks.reduce((acc, t) => {
    if (t && !acc.includes(t.pool)) acc.push(t.pool);
    return acc;
  }, []);


  // Remove tasks from all states collections, except `finished`
  //
  let query = this.queue.__redis__.multi()
    .srem(`${prefix}waiting`, cancelIDs.concat([ this.id ]))
    .srem(`${prefix}idle`, cancelIDs.concat([ this.id ]))
    .zrem(`${prefix}locked`, cancelIDs.concat([ this.id ]))
    .zrem(`${prefix}restart`, cancelIDs.concat([ this.id ]));

  pools.forEach(pool => query.srem(`${prefix}${pool}:startable`, cancelIDs.concat([ this.id ])));

  // Calculate number of commands before transactions, to get results later
  let shift = 4 + pools.length;

  // Add transaction for this task:
  //
  // - move this task to `finished`
  // - update state
  // - save child error
  // - send command with error to parent
  //
  let transaction = {
    if: [
      [ 1, '==', [ 'zrem', `${prefix}${this.pool}:commands_locked`, command.toJSON() ] ],
      [ JSON.stringify('finished'), '!=', [ 'hget', `${prefix}${this.id}`, 'state' ] ]
    ],
    exec: [
      [ 'zadd', `${prefix}finished`, time + this.removeDelay, this.id ],
      [ 'hset', `${prefix}${this.id}`, 'state', JSON.stringify('finished') ],
      [ 'hset', `${prefix}${this.id}`, 'error', JSON.stringify(command.data.error) ]
    ]
  };

  if (this.parent) {
    // Send command with error to parent
    transaction.exec.push([ 'zadd', `${prefix}${this.parent_pool}:commands`, time, Command.fromObject({
      to:     this.parent,
      to_uid: this.parent_uid,
      type:   'error',
      data:   command.data
    }).toJSON() ]);
  }

  query.transaction(JSON.stringify(transaction));


  // Add transactions for children:
  //
  // - move tasks to `finished`
  // - update their state
  // - save error
  //
  let state = JSON.stringify('finished');
  let error = JSON.stringify({ code: 'CANCELED', name: 'Error', message: 'idoit error: task canceled' });

  cancelIDs.forEach((id, i) => {
    // If task already deleted/cancelled - skip
    if (!rawTasks[i] || rawTasks[i].state === 'finished') return;

    let removeDeadline = time + rawTasks[i].removeDelay;

    query.transaction(
      JSON.stringify({
        if: [
          [ state, '!=', [ 'hget', `${prefix}${id}`, 'state' ] ]
        ],
        exec: [
          [ 'zadd', `${prefix}finished`, removeDeadline, id ],
          [ 'hset', `${prefix}${id}`, 'state', state ],
          [ 'hset', `${prefix}${id}`, 'error', error ]
        ]
      })
    );
  });


  let res = await query.exec();

  if (res[shift][1]) { // transaction result
    let eventData = { id: this.id, uid: this.uid };

    this.queue.emit('task:end', eventData);
    this.queue.emit(`task:end:${this.id}`, eventData);
  }
};


TaskTemplate.prototype.__fillCancelInfo__ = async function () {
  return [ this ];
};


// Fields for `.toObject()/.fromObject()`
//
TaskTemplate.serializableFields = [
  'name',
  'id',
  'pool',
  'uid',
  'retries',
  'total',
  'progress',
  'state',
  'parent',
  'parent_uid',
  'parent_pool',
  'args',
  'result',
  'error',
  // keep it to access without task class cast on cancel/error
  'removeDelay'
];


// Serialize task to plain object
//
TaskTemplate.prototype.toObject = function () {
  return this.__serializableFields.reduce((obj, key) => {
    if (key in this) obj[key] = this[key];
    return obj;
  }, {});
};


// Unserialize task from plain object
//
TaskTemplate.prototype.fromObject = function (obj) {
  this.__serializableFields.forEach(key => {
    if (obj.hasOwnProperty(key)) this[key] = obj[key];
  });
};


// Helper to fill parent references
//
TaskTemplate.prototype.setParent = function (parentTask) {
  this.parent      = parentTask.id;
  this.parent_uid  = parentTask.uid;
  this.parent_pool = parentTask.pool;
};


// Task class factory
//
TaskTemplate.extend = function (options) {
  if (typeof options.process !== 'function') {
    throw new Error('idoit error: "process" should be a function');
  }

  class T extends this {}

  Object.assign(T.prototype, options);

  return T;
};


module.exports = TaskTemplate;
